rust_icu_unum/src/lib.rs

```diff
@@ -443,6 +675,42 @@ impl<'a, T> Iterator for UFieldPositionIterator<'a, T> {
     }
 }
 
+/// Gets an iterator over all available formatting locales.
+///
+/// Implements `unum_getAvailable`. Since 0.3.1.
+pub fn available_iter() -> UnumIter {
+    let max = unsafe { versioned_function!(unum_countAvailable)() } as usize;
+    UnumIter { max, next: 0 }
+}
+
+/// An iterator returned by `available_iter()`, containing the string representation of locales for
+/// which formatting is available.
+pub struct UnumIter {
+    /// The total number of elements that this iterator can yield.
+    max: usize,
+    /// The next element index that can be yielded.
+    next: usize,
+}
+
+impl Iterator for UnumIter {
+    type Item = &'static str;
+
+    fn next(&mut self) -> Option<&'static str> {
+        if self.max == 0 {
+            return None;
+        }
+        if self.next >= self.max {
+            return None;
+        }
+        let c = unsafe {
+            let cptr = versioned_function!(unum_getAvailable)(self.next as i32);
+            std::ffi::CStr::from_ptr(cptr)
+        };
+        self.next = self.next + 1;
+        Some(c.to_str().expect("can be converted to str"))
```

Comment:
It might be worth commenting why we're confident that `'static` is the correct lifetime here.

Reply:
*Excellent* point!  

I guess, there are no guarantees.  

When writing this, I thought that this call would be returning compiled-in values (hence `'static`), but then realized this is not true when locale data is loaded from a file, and is *especially* not true if the data is reloaded during the lifetime of the iterator.

And since there are no global objects to hang the lifetime off of, this has to return `Option<String>` instead.

There are other gotchas remaining, such as any data invalidation that happens while the iterator is live.  Which I don't have a good idea how to remove.